src/main/java/org/prebid/server/auction/requestfactory/Ortb2ImplicitParametersResolver.java

```diff
@@ -427,14 +445,29 @@ private Source populateSource(Source source, ExtRequest extRequest) {
         return null;
     }
 
-    private String populateSourceTid(String tid) {
-        if (StringUtils.isNotEmpty(tid)) {
+    private static String populateTidValue(String tid,
+                                           boolean generateBidRequestId,
+                                           String endpoint,
+                                           boolean hasStoredBidRequest,
+                                           IdGenerator tidGenerator) {
+
+        final boolean containsTidMacro = StringUtils.containsIgnoreCase(tid, OVERRIDE_SOURCE_ID_TEMPLATE);
+        if (StringUtils.isNotBlank(tid)
+                && !containsTidMacro
+                && !(generateBidRequestId
+                && (org.prebid.server.model.Endpoint.openrtb2_amp.value().equals(endpoint)
+                || hasStoredBidRequest))) {
             return null;
         }
 
-        final String generatedId = sourceIdGenerator.generateId();
+        final String generatedId = tidGenerator.generateId();
 
-        return StringUtils.isNotEmpty(generatedId) ? generatedId : null;
+        return StringUtils.isBlank(tid)
+                || (generateBidRequestId
+                && (org.prebid.server.model.Endpoint.openrtb2_amp.value().equals(endpoint) || hasStoredBidRequest))
+                && !containsTidMacro
+                ? generatedId
+                : StringUtils.replaceIgnoreCase(tid, OVERRIDE_SOURCE_ID_TEMPLATE, generatedId);
     }
```

Comment:
I think that `Endpoint.openrtb2_amp.value().equals(endpoint)` is redundant. We can always set `hasStoredBidRequest` as `true` for `amp`.